# nf-core/ssds: Output

## Introduction

This document describes the output produced by the SSDS pipeline. The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview
	
The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

* [fastqc](#fastqc) - Raw read QC
* [bam](#bedbam") - ssDNA / dsDNA -derived fragments in BAM format
* [bed](#bedbam) - ssDNA / dsDNA -derived fragments in BED format
* [bigwig](#bigwig) - Genome coverage tracks
* [reports](#reports) - Various reports generated by the pipeline
* [multiqc](#multiqc) - Aggregate report describing results and QC from the whole pipeline
* [pipeline info](#pipeline-information) - Report metrics generated during the workflow execution

### fastqc

<details markdown="1">
<summary>Output files</summary>

* `fastqc/`
    * `*_fastqc.html`: FastQC report containing quality metrics.
    * `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

</details>

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences. For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

![MultiQC - FastQC sequence counts plot](images/mqc_fastqc_counts.png)

![MultiQC - FastQC mean quality scores plot](images/mqc_fastqc_quality.png)

![MultiQC - FastQC adapter content plot](images/mqc_fastqc_adapter.png)

> **NB:** The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.

### bedbam

<details markdown="1">
<summary>Output files</summary>

* `bam/`
    * `<sample>.ssDNA.bam`: a BAM file containing ssDNA-derived reads.
	* `<sample>.ssDNA.bam`: an index for a BAM file containing ssDNA-derived reads.
	* `<sample>.<type>.bam`: BAM files containing reads with other type classifications.
    * `<sample>.<type>.bam`: indexes for BAM files containing reads with other type classifications.
    * `<sample>.unclassified.bam`: a BAM file containing unclassified reads.
	* `<sample>.unclassified.bam`: an index for a BAM file containing unclassified reads.    

* `bed/`
    * `<sample>.ssDNA.bed`: a BED file containing ssDNA-derived fragments.
    * `<sample>.<type>.bed`: BED files containing fragments with other type classifications.
    * `<sample>.unclassified.bam`: a BED file containing unclassified fragments.

</details>

- [BWA](https://github.com/lh3/bwa) is used to align reads to the genome. 
- [Picard](https://broadinstitute.github.io/picard/) is used to sort the aligned BAM files and to Mark Duplicate reads. 
- [Samtools](http://www.htslib.org/) is used to index the sorted BAM files. 

- The [SSDS toolkit](https://github.com/kevbrick/ssds_pipeline_accessory_scripts) parses the aligned BAM file from an SSDS experiment. SSDS allows sequencing of single stranded-DNA (ssDNA) because of a unique end-repair process that allows sequencing adapters to ligate to ss-DNA. This end-repair creates characteristic structures that allow ss-DNA-derived reads to be identified. The [SSDS toolkit](https://github.com/kevbrick/ssds_pipeline_accessory_scripts) classifies reads and outputs each type to a paired-end BAM file and to a BED file of fragments (useful for peak calling).

	In "ss" mode (default), two fragment classes are defined:   
	1. ssDNA : Read pairs with a signature unambiguously derived from ss-DNA. For all published SSDS experiments, these have been the only reads used.
	
	2. unclassified
	
	In "all" mode, five fragment classes are defined (NOTE that only type 1 - ssDNA - is unambiguous): 
	1. ssDNA : Read pairs with a signature unambiguously derived from ss-DNA. For all published SSDS experiments, these have been the only reads used.
	
	2. type-2 ssDNA : Read pairs with a signature likely, but not unambiguously derived from ss-DNA. 

	3. high-confidence dsDNA : Read pairs with a signature likely, but not unambiguously derived from ds-DNA. 

	4. low-confidence dsDNA : Read pairs with a signature indicative of ds-DNA, but possible derived from ssDNA.  

	5. unclassified

### bigwig

<details markdown="1">
<summary>Output files</summary>

* `bigwig/`
    * `<sample>.<read_type>.FWD.bigwig`: Genome coverage of reads that map to the forward (top) strand. 
    * `<sample>.<read_type>.REV.bigwig`: Genome coverage of reads that map to the reverse (bottom) strand. 
    * `<sample>.<read_type>.TOT.bigwig`: Genome coverage of all reads. 
    * `<sample>.<read_type>.FR.bigwig`: Genome coverage of the log2(fwd/reverse) read ratio. 
    
</details>


- [BEDtools](https://bedtools.readthedocs.io/en/latest/) is used to divide the genome into equal intervals and to count the number of reads per window. 
- [UCSC tools (BedGraphToBigWig)](https://hgdownload.soe.ucsc.edu/downloads.html#utilities_downloads) is used to convert bedgraphs to bigwig format. 

> **Note:** Coverage tracks will not be generated for files with very few reads.

### reports

<details markdown="1">
<summary>Output files</summary>

* `reports/`
    * `<sample_id>.*`: various reports used by multiqc.

</details>

Raw report files are output here.
### multiqc

<details markdown="1">
<summary>Output files</summary>

* `multiqc/`
    * `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
    * `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
    * `multiqc_plots/`: directory containing static images from the report in various formats.

</details>

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarising all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

Results generated by MultiQC collate pipeline QC from supported tools e.g. FastQC. The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability. For more information about how to use MultiQC reports, see <http://multiqc.info>.

The SSDS module for MultiQC generates three major reports: 
1. Sample content report - what proportion of reads are derived from each class of molecule
2. Fragment properties report - Distributions of fragment properties used for inferring read-pair source
3. SPoT report - (Optional) - if intervals are provided to the pipeline in conf/spot_intervals.conf, then this will show the percent of reads of each type that occur in those intervals. 


### Pipeline information

<details markdown="1">
<summary>Output files</summary>

* `pipeline_info/`
    * Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
    * Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
    * Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
